syft/frameworks/torch/tensors/interpreters/replicated_shared.py

```diff
@@ -79,63 +93,152 @@ def __map_modular_to_real(self, mod_number):
         players = self.get_players()
         shares_map = self.get_shares_map()
-        plain_text = torch.tensor(plain_text).send(players[0])
+        plain_text = torch.tensor(plain_text).long().send(players[0])
         shares_map[players[0]] = (
             operator(shares_map[players[0]][0], plain_text),
             shares_map[players[0]][1],
         )
-        return syft.ReplicatedSharingTensor(shares_map)
+        return ReplicatedSharingTensor().__set_shares_map(shares_map)
+
+    def mul(self, value):
+        return self.__switch_public_private(value, self.public_mul, self.private_mul)
+
+    def public_mul(self, plain_text):
+        return self.__public_multiplication_operation(plain_text, mul)
+
+    def private_mul(self, secret):
+        return self.__private_multiplication_operation(secret, mul)
+
+    __mul__ = mul
+
+    def matmul(self, value):
+        return self.__switch_public_private(value, self.public_matmul, self.private_matmul)
+
+    def public_matmul(self, plain_text):
+        return self.__public_multiplication_operation(plain_text, torch.matmul)
+
+    def private_matmul(self, secret):
+        return self.__private_multiplication_operation(secret, torch.matmul)
+
+    __matmul__ = matmul
+
+    def __public_multiplication_operation(self, plain_text, operator):
+        players = self.get_players()
+        plain_text_map = {
+            player: torch.tensor(plain_text).long().send(player) for player in players
+        }
+        shares_map = self.get_shares_map()
+        for player in players:
+            shares_map[player] = (
+                operator(shares_map[player][0], plain_text_map[player]),
+                operator(shares_map[player][1], plain_text_map[player]),
+            )
+        return ReplicatedSharingTensor().__set_shares_map(shares_map)
+
+    def __private_multiplication_operation(self, secret, operator):
+        x, y = self.get_shares_map(), secret.get_shares_map()
+        players = self.get_players()
+        z = [
+            operator(x[player][0], y[player][0])
+            + operator(x[player][1], y[player][0])
+            + operator(x[player][0], y[player][1])
+            for player in players
+        ]
+        z = self.__add_noise(z)
+        z = self.__reshare(z, players)
+        return ReplicatedSharingTensor().__set_shares_map(z)
+
+    @staticmethod
+    def __add_noise(shares):
+        noisy_shares = [share + gen_alpha_3of3(share.location).wrap() for share in shares]
+        return noisy_shares
+
+    @staticmethod
+    def __reshare(shares, workers):
+        """convert 3-out-of-3 secret sharing: {player i : share i}
+          to 2-out-of-3 sharing: {player i : (share i, share i+1)}  """
+        shares_map = {}
+        for i in range(len(shares)):
+            pointer = shares[(i + 1) % len(shares)].copy().move(workers[i])
+            shares_map[workers[i]] = (shares[i], pointer)
+        return shares_map
+
+    def conv2d(self, image, padding=0):
```

Comment:
Q: Should we have this in the ```RSS``` or in the ```Falcon``` class?

Reply:
hmm, it could be implemented both ways, but the paper described it as an operation of the sharing scheme, so let's stick to the paper for now, it can be changed easily later if it is inconvenient